src/CMakeLists.txt

```diff
@@ -1,97 +1,78 @@
 if(SECP256K1_ASM STREQUAL "arm")
-  add_library(common OBJECT
+  add_library(secp_asm_arm OBJECT EXCLUDE_FROM_ALL)
+  target_sources(secp_asm_arm PUBLIC
     asm/field_10x26_arm.s
   )
-  set(common_obj "$<TARGET_OBJECTS:common>")
-else()
-  set(common_obj "")
+  target_sources(secp256k1 PRIVATE $<TARGET_OBJECTS:secp_asm_arm>)
+  target_link_libraries(secp_asm INTERFACE secp_asm_arm)
 endif()
 
-add_library(precomputed OBJECT
-  precomputed_ecmult.c
-  precomputed_ecmult_gen.c
-)
-set(internal_obj "$<TARGET_OBJECTS:precomputed>" "${common_obj}")
+#Define our export symbol only for Win32 and only for shared libs.
+if (WIN32)
+  set_target_properties(secp256k1 PROPERTIES DEFINE_SYMBOL "DLL_EXPORT")
+endif()
 
-add_library(secp256k1 SHARED EXCLUDE_FROM_ALL
-  secp256k1.c
-  ${internal_obj}
-)
-target_include_directories(secp256k1 INTERFACE
+# Object libs don't know if they're being built for a shared or static lib.
+# Grab the PIC property from secp256k1 which knows.
+get_target_property(use_pic secp256k1 POSITION_INDEPENDENT_CODE)
+set_target_properties(secp_precomputed_objs PROPERTIES POSITION_INDEPENDENT_CODE ${use_pic})
+
+target_include_directories(secp256k1 PUBLIC
   $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>
 )
-target_compile_definitions(secp256k1 PRIVATE
-  $<$<PLATFORM_ID:Windows>:DLL_EXPORT>
-)
 set_target_properties(secp256k1 PROPERTIES
   VERSION "${${PROJECT_NAME}_LIB_VERSION_CURRENT}.${${PROJECT_NAME}_LIB_VERSION_AGE}.${${PROJECT_NAME}_LIB_VERSION_REVISION}"
   SOVERSION "${${PROJECT_NAME}_LIB_VERSION_CURRENT}"
 )
-if(SECP256K1_BUILD_SHARED)
-  get_target_property(use_pic secp256k1 POSITION_INDEPENDENT_CODE)
-  set_target_properties(precomputed PROPERTIES POSITION_INDEPENDENT_CODE ${use_pic})
-  set_target_properties(secp256k1 PROPERTIES EXCLUDE_FROM_ALL FALSE)
-  list(APPEND ${PROJECT_NAME}_installables secp256k1)
-endif()
-
-add_library(secp256k1_static STATIC EXCLUDE_FROM_ALL
-  secp256k1.c
-  ${internal_obj}
-)
-target_include_directories(secp256k1_static INTERFACE
-  $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>
-)
-if(NOT MSVC)
-  set_target_properties(secp256k1_static PROPERTIES
-    OUTPUT_NAME secp256k1
-  )
-endif()
-if(SECP256K1_BUILD_STATIC)
-  set_target_properties(secp256k1_static PROPERTIES EXCLUDE_FROM_ALL FALSE)
-  list(APPEND ${PROJECT_NAME}_installables secp256k1_static)
-endif()
-
-add_library(link_library INTERFACE)
-if(SECP256K1_BUILD_SHARED)
-  target_link_libraries(link_library INTERFACE secp256k1)
-elseif(SECP256K1_BUILD_STATIC)
-  target_link_libraries(link_library INTERFACE secp256k1_static)
-endif()
 
 if(SECP256K1_BUILD_BENCHMARK)
   add_executable(bench bench.c)
-  target_link_libraries(bench link_library)
-  add_executable(bench_internal bench_internal.c ${internal_obj})
-  add_executable(bench_ecmult bench_ecmult.c ${internal_obj})
+  target_link_libraries(bench secp256k1)
+  add_executable(bench_internal bench_internal.c)
+  target_link_libraries(bench_internal secp_precomputed_objs secp_asm)
```

Comment:
> Would it make sense (if it's possible) to define a library/target/whatever that combines secp256k1_precomputed_objs and secp_asm?

I was thinking about the same :)
The diff as follows can help:
```diff
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -10,14 +10,12 @@ add_library(secp_precomputed_objs OBJECT EXCLUDE_FROM_ALL
 # from being exported.
 add_library(secp256k1 secp256k1.c $<TARGET_OBJECTS:secp_precomputed_objs>)
 
-add_library(secp_asm INTERFACE)
 if(SECP256K1_ASM STREQUAL "arm")
   add_library(secp_asm_arm OBJECT EXCLUDE_FROM_ALL)
   target_sources(secp_asm_arm PUBLIC
     asm/field_10x26_arm.s
   )
   target_sources(secp256k1 PRIVATE $<TARGET_OBJECTS:secp_asm_arm>)
-  target_link_libraries(secp_asm INTERFACE secp_asm_arm)
 endif()
 
 #Define our export symbol only for Win32 and only for shared libs.
@@ -38,23 +36,29 @@ set_target_properties(secp256k1 PROPERTIES
   SOVERSION "${${PROJECT_NAME}_LIB_VERSION_CURRENT}"
 )
 
+add_library(secp256k1_internal INTERFACE)
+set_property(TARGET secp256k1_internal PROPERTY INTERFACE_SOURCES
+  $<TARGET_OBJECTS:secp_precomputed_objs>
+  $<$<TARGET_EXISTS:secp_asm_arm>:$<TARGET_OBJECTS:secp_asm_arm>>
+)
+
 if(SECP256K1_BUILD_BENCHMARK)
   add_executable(bench bench.c)
   target_link_libraries(bench secp256k1)
   add_executable(bench_internal bench_internal.c)
-  target_link_libraries(bench_internal secp_precomputed_objs secp_asm)
+  target_link_libraries(bench_internal secp256k1_internal)
   add_executable(bench_ecmult bench_ecmult.c)
-  target_link_libraries(bench_ecmult secp_precomputed_objs secp_asm)
+  target_link_libraries(bench_ecmult secp256k1_internal)
 endif()
 
 if(SECP256K1_BUILD_TESTS)
   add_executable(noverify_tests tests.c)
-  target_link_libraries(noverify_tests secp_precomputed_objs secp_asm)
+  target_link_libraries(noverify_tests secp256k1_internal)
   add_test(noverify_tests noverify_tests)
   if(NOT CMAKE_BUILD_TYPE STREQUAL "Coverage")
     add_executable(tests tests.c)
     target_compile_definitions(tests PRIVATE VERIFY)
-    target_link_libraries(tests secp_precomputed_objs secp_asm)
+    target_link_libraries(tests secp256k1_internal)
     add_test(tests tests)
   endif()
 endif()
@@ -62,7 +66,7 @@ endif()
 if(SECP256K1_BUILD_EXHAUSTIVE_TESTS)
   # Note: do not include secp_precomputed_objs in exhaustive_tests (it uses runtime-generated tables).
   add_executable(exhaustive_tests tests_exhaustive.c)
-  target_link_libraries(exhaustive_tests secp_asm)
+  target_link_libraries(exhaustive_tests $<TARGET_NAME_IF_EXISTS:secp_asm_arm>)
   target_compile_definitions(exhaustive_tests PRIVATE $<$<NOT:$<CONFIG:Coverage>>:VERIFY>)
   add_test(exhaustive_tests exhaustive_tests)
 endif()
```

Reply:
> Would it make sense (if it's possible) to define a library/target/whatever that combines secp256k1_precomputed_objs and secp_asm?

I didn't combine them here because I wasn't sure if the combo of `secp256k1_precomputed_objs` and `secp_asm` made any logical sense worth defining, as opposed to just listing them explicitly where needed.

Happy to take a patch like @hebasto's above.

> Also naming... can we be more consistent than having secp256k1 in one and secp in the other?

Roger.

Reply:
Imo the complexity here isn't worth it and I'd rather not take this change unless anyone feels strongly about it.